### ISSUE TITLE
programmatically setting baseUrl in java / retrofit2

### ISSUE THREAD

**Issue by haroldcarr** [swagger-api/swagger-codegen#3891]
##### Description

Using Java / retrofit2 client generation.

It generates `ApiClient` where `baseUrl` is locally set in `createDefaultAdapter` and passed to `Retrofit.Builder`.

I want to use this same generated code, but with different baseUrls programmatically.

That does not seem to be possible.

##### Swagger-codegen version

2.2.2-SNAPSHOT

##### Command line used for generation

    java -jar ${swaggerCodegenJar} generate \
         -l java \
         --library retrofit2 \
         --group-id TO-BE-DELETED \
         --artifact-id java-retrofit2-client \
         --api-package api \
         --invoker-package invoker \
         --model-package model \
         -i DocService-folders-swagger.json \
         -o /tmp

##### Related issues

https://github.com/swagger-api/swagger-codegen/issues/2823

##### Suggest a Fix

Rather than have a bunch of different constructors for `ApiClient`, use a "fluent/builder" style that includes `baseUrl` as one of its methods.

**Comment by wing328** [comment]
@haroldcarr would the following help?

https://github.com/swagger-api/swagger-codegen/issues/2769#issuecomment-217674257

**Comment by haroldcarr** [comment]
@wing328 It seems like that should work.  I will try it tomorrow morning and report back here.  Thanks.

**Comment by haroldcarr** [comment]
@wing328 that worked - except the method is different:

    apiClient.getAdapterBuilder().baseUrl(URL);

**Comment by wing328** [comment]
@haroldcarr thanks for the feedback. I think what we'll do is to add this to the auto-generated documentation (MD) so that users will have a better idea on how to set the base URL.

**Comment by haroldcarr** [comment]
I assume you will close the issue once the doc is updated?

**Comment by natami** [comment]
There is still a problem with baseurl.

When dealing with OAuth in the api client.

See the generated code:
        for(String authName : authNames) { 
            Interceptor auth;
            if (authName == "appFlowSecurityDef") { 
                auth = new OAuth(OAuthFlow.application, "http://localhost:9090/oauth/token", "http://localhost:9090/oauth/token", "users:write");
            } else if (authName == "iotServerFlowSecurityDef") { 
                auth = new OAuth(OAuthFlow.application, "http://localhost:9090/oauth/token", "http://localhost:9090/oauth/token", "users:verify");
            } else if (authName == "userFlowSecurityDef") { 
                auth = new OAuth(OAuthFlow.password, "http://localhost:9090/oauth/token", "http://localhost:9090/oauth/token", "users:identify, data:statistics");
            } else {
                throw new RuntimeException("auth name \"" + authName + "\" not found in available auth names");
            }
            addAuthorization(authName, auth);
        }

Is there a way to change the authorizationUrl and tokenUrl?

- Maybe this could be based on the baseurl path?

**Comment by cbornet** [comment]
Yes. You can add your own authorization to the authorization list
```java
addAuthorization("my_auth", new OAuth(OAuthFlow.password, "http://www.foo.com:9090/oauth/token", "http://www.foo.com:9090/oauth/token", "users:identify, data:statistics"))
```
It can't be based on baseUrl because some (many) systems have a different endpoint/server to get the token.
The authorizationUrl  and tokenUrl used are taken from securityDefinition field in the swagger spec so you can modify them there.

**Comment by wshreek** [comment]
Is this issue only tackling for java client?

We are using typescript client and soon we will start using C# client. We definitely would like to have configurable endpoint, coz every developer on the team would like to point to their own localhost. 

Please let me know if there's a solution/workaround for typescript & C#.

**Comment by wing328** [comment]
@wshreek for C#, you can change the base URL via the default configuration.

For typescript, are you referring to Node, Fetch, Angular, Angular2? I believe there's something similar you can do with ApiClient or Configuration class to customize the base URL.

**Comment by wing328** [comment]
@wshreek if you still need help customizing the base URL for other API clients, please open a new issue for tracking.
